console/src/components/Services/Data/Function/customFunctionReducer.js

```diff
@@ -0,0 +1,365 @@
+const UNTRACK_CUSTOM_FUNCTION_FAIL =
+  '@customFunction/UNTRACK_CUSTOM_FUNCTION_FAIL';
+
+/* */
+
+const makeRequest = (
+  upQueries,
+  downQueries,
+  migrationName,
+  customOnSuccess,
+  customOnError,
+  requestMsg,
+  successMsg,
+  errorMsg
+) => {
+  return (dispatch, getState) => {
+    const upQuery = {
+      type: 'bulk',
+      args: upQueries,
+    };
+
+    const downQuery = {
+      type: 'bulk',
+      args: downQueries,
+    };
+
+    const migrationBody = {
+      name: migrationName,
+      up: upQuery.args,
+      down: downQuery.args,
+    };
+
+    const currMigrationMode = getState().main.migrationMode;
+
+    const migrateUrl = returnMigrateUrl(currMigrationMode);
+
+    let finalReqBody;
+    if (globals.consoleMode === SERVER_CONSOLE_MODE) {
+      finalReqBody = upQuery;
+    } else if (globals.consoleMode === 'cli') {
+      finalReqBody = migrationBody;
+    }
+    const url = migrateUrl;
+    const options = {
+      method: 'POST',
+      credentials: globalCookiePolicy,
+      headers: dataHeaders(getState),
+      body: JSON.stringify(finalReqBody),
+    };
+
+    const onSuccess = data => {
+      if (globals.consoleMode === 'cli') {
+        dispatch(loadMigrationStatus()); // don't call for server mode
+      }
+      // dispatch(loadTriggers());
+      if (successMsg) {
+        dispatch(showSuccessNotification(successMsg));
+      }
+      customOnSuccess(data);
+    };
+
+    const onError = err => {
+      dispatch(handleMigrationErrors(errorMsg, err));
+      customOnError(err);
+    };
+
+    dispatch(showSuccessNotification(requestMsg));
+    return dispatch(requestAction(url, options)).then(onSuccess, onError);
+  };
+};
+
+/* Action creators */
+const fetchCustomFunction = (functionName, schema) => {
+  return (dispatch, getState) => {
+    const url = Endpoints.getSchema;
+    const fetchCustomFunctionQuery = {
+      type: 'select',
+      args: {
+        table: {
+          name: 'hdb_function',
+          schema: 'hdb_catalog',
+        },
+        columns: ['*'],
+        where: {
+          function_schema: schema,
+          function_name: functionName,
+        },
+      },
+    };
+    const fetchCustomFunctionDefinition = {
+      type: 'select',
+      args: {
+        table: {
+          name: 'hdb_function_agg',
+          schema: 'hdb_catalog',
+        },
+        columns: ['*'],
+        where: {
+          function_schema: schema,
+          function_name: functionName,
+        },
+      },
+    };
+
+    const bulkQuery = {
+      type: 'bulk',
+      args: [fetchCustomFunctionQuery, fetchCustomFunctionDefinition],
+    };
+    const options = {
+      credentials: globalCookiePolicy,
+      method: 'POST',
+      headers: dataHeaders(getState),
+      body: JSON.stringify(bulkQuery),
+    };
+    dispatch({ type: FETCHING_INDIV_CUSTOM_FUNCTION });
+    return dispatch(requestAction(url, options)).then(
+      data => {
+        if (data[0].length > 0 && data[1].length > 0) {
```

Comment:
What if data is an empty array? It will throw an exception.

Reply:
Its a bulk query. I am guessing the base value will be `[[],[]]`.